scalability and 2-10x performance improvements

bobby-stripe opened the pull request:
There are 4 major pieces to this PR:
* use a `sync.Pool` of `rand.Rand`s rather than a mutex to allow better scalability
* use regular `map`s behind an `atomic.Value` for `flags` and `defaultTags` rather than a `sync.Map`.  When used, the maps are read-only, and the config data itself is read-mostly, so avoiding synchronization here is a big win
* Rather than `randFunc` use a `randFloater` (happy for better name here) interface: using the `g.rand` method as a `randFunc` was escaping to the heap and causing an allocation every time `Enabled` was fractionally on.
* If `Enabled` is called with an empty `properties`, avoid allocating a new map

benchmarks:

```
benchmark                          old ns/op     new ns/op     delta
BenchmarkEnabled/csv/50pct         395           171           -56.70%
BenchmarkEnabled/csv/50pct-2       144           48.8          -66.18%
BenchmarkEnabled/csv/50pct-4       126           21.3          -83.11%
BenchmarkEnabled/csv/50pct-8       148           14.1          -90.46%
BenchmarkEnabled/csv/on            262           164           -37.21%
BenchmarkEnabled/csv/on-2          35.7          16.4          -53.98%
BenchmarkEnabled/csv/on-4          15.0          6.30          -57.97%
BenchmarkEnabled/csv/on-8          9.49          4.25          -55.19%
BenchmarkEnabled/json2/50pct       1502          568           -62.15%
BenchmarkEnabled/json2/50pct-2     137           58.5          -57.36%
BenchmarkEnabled/json2/50pct-4     128           23.1          -82.03%
BenchmarkEnabled/json2/50pct-8     175           14.4          -91.81%
BenchmarkEnabled/json2/on          330           117           -64.61%
BenchmarkEnabled/json2/on-2        38.7          13.5          -65.09%
BenchmarkEnabled/json2/on-4        13.5          6.67          -50.48%
BenchmarkEnabled/json2/on-8        9.08          4.13          -54.51%

benchmark                          old allocs     new allocs     delta
BenchmarkEnabled/csv/50pct         2              0              -100.00%
BenchmarkEnabled/csv/50pct-2       2              0              -100.00%
BenchmarkEnabled/csv/50pct-4       2              0              -100.00%
BenchmarkEnabled/csv/50pct-8       2              0              -100.00%
BenchmarkEnabled/csv/on            0              0              +0.00%
BenchmarkEnabled/csv/on-2          0              0              +0.00%
BenchmarkEnabled/csv/on-4          0              0              +0.00%
BenchmarkEnabled/csv/on-8          0              0              +0.00%
BenchmarkEnabled/json2/50pct       5              1              -80.00%
BenchmarkEnabled/json2/50pct-2     2              0              -100.00%
BenchmarkEnabled/json2/50pct-4     2              0              -100.00%
BenchmarkEnabled/json2/50pct-8     2              0              -100.00%
BenchmarkEnabled/json2/on          0              0              +0.00%
BenchmarkEnabled/json2/on-2        0              0              +0.00%
BenchmarkEnabled/json2/on-4        0              0              +0.00%
BenchmarkEnabled/json2/on-8        0              0              +0.00%

benchmark                          old bytes     new bytes     delta
BenchmarkEnabled/csv/50pct         150           37            -75.33%
BenchmarkEnabled/csv/50pct-2       104           10            -90.38%
BenchmarkEnabled/csv/50pct-4       98            2             -97.96%
BenchmarkEnabled/csv/50pct-8       102           1             -99.02%
BenchmarkEnabled/csv/on            74            51            -31.08%
BenchmarkEnabled/csv/on-2          9             4             -55.56%
BenchmarkEnabled/csv/on-4          3             1             -66.67%
BenchmarkEnabled/csv/on-8          1             0             -100.00%
BenchmarkEnabled/json2/50pct       319           100           -68.65%
BenchmarkEnabled/json2/50pct-2     82            9             -89.02%
BenchmarkEnabled/json2/50pct-4     84            3             -96.43%
BenchmarkEnabled/json2/50pct-8     90            0             -100.00%
BenchmarkEnabled/json2/on          53            18            -66.04%
BenchmarkEnabled/json2/on-2        6             1             -83.33%
BenchmarkEnabled/json2/on-4        0             0             +0.00%
BenchmarkEnabled/json2/on-8        0             0             +0.00%
```

vasi-stripe commented on goforit_test.go:
```diff
@@ -251,9 +251,11 @@ func BenchmarkEnabled(b *testing.B) {
-				for i := 0; i < b.N; i++ {
-					_ = g.Enabled(context.Background(), flag.flag, nil)
-				}
+				b.RunParallel(func(pb *testing.PB) {
```
Hmm, maybe we want both the old and new versions here? We care about both latency _and_ throughput.

bobby-stripe replied:
@vasi-stripe I think the old version is equivalent to running the new version with `-cpu 1` on the command line (for running a single instance at once).  I can add the original form back if you think it would be less confusing (or if there is a difference I'm not seeing here?)

vasi-stripe replied:
I'm not really understanding why the old version benches differently at -cpu 1 vs -cpu 2! But it does look like this improves even the single-cpu results, so that's good enough for me.

bobby-stripe replied:
@vasi-stripe writing here for completeness:  I think whats happening is that `-cpu` internally sets `GOMAXPROCS` for each test.  for the `-cpu 1` case, we end up with contention on that single Go `proc`/processor: a single instance of the benchmark is running on it, but so is the flags backend goroutine which is calling `g.RefreshFlags()` every `10*time.Microsecond`. The non-`RunParallel` tests used default `GOMAXPROCS`, so the single goroutine running the benchmark was almost always free to run in parallel with the backend refreshing the tags.

bobby-stripe commented:
r-codeowners?

bobby-stripe commented:
re-r? @vasi-stripe 

I added one new commit (rest is unchanged but rebased on `origin/master`) that matches the old behavior of `Get` on the flag map returning a struct instead of a pointer to a struct.  Looking at the code before merging, `Enabled` made some assumptions that even if the flag didn't exist the holder was non-nil.  I added a test for that too (that failed before the new PR).

Performance is unchanged:

```
$ benchstat orig.txt proposed.txt
name                   old time/op    new time/op    delta
Enabled/csv/50pct         179ns ± 1%     184ns ± 2%   ~     (p=0.333 n=2+2)
Enabled/csv/50pct-2      46.0ns ± 1%    48.6ns ± 2%   ~     (p=0.333 n=2+2)
Enabled/csv/50pct-4      21.0ns ± 4%    22.4ns ± 3%   ~     (p=0.667 n=2+2)
Enabled/csv/on            153ns ±46%      80ns ± 3%   ~     (p=0.333 n=2+2)
Enabled/csv/on-2         15.5ns ± 4%    13.8ns ± 2%   ~     (p=0.333 n=2+2)
Enabled/csv/on-4         6.46ns ± 2%    6.31ns ± 5%   ~     (p=1.000 n=2+2)
Enabled/json2/50pct       602ns ±27%     686ns ±13%   ~     (p=0.667 n=2+2)
Enabled/json2/50pct-2    63.1ns ± 2%    54.7ns ±11%   ~     (p=0.333 n=2+2)
Enabled/json2/50pct-4    21.1ns ± 2%    21.7ns ± 1%   ~     (p=0.667 n=2+2)
Enabled/json2/on          153ns ± 1%     206ns ± 5%   ~     (p=0.333 n=2+2)
Enabled/json2/on-2       15.3ns ±12%    16.3ns ±11%   ~     (p=0.667 n=2+2)
Enabled/json2/on-4       6.34ns ± 6%    6.14ns ± 1%   ~     (p=1.000 n=2+2)

name                   old alloc/op   new alloc/op   delta
Enabled/csv/50pct         38.0B ± 5%     41.5B ± 4%   ~     (p=0.667 n=2+2)
Enabled/csv/50pct-2       7.00B ±14%     9.50B ± 5%   ~     (p=0.333 n=2+2)
Enabled/csv/50pct-4       1.50B ±33%     3.00B ±33%   ~     (p=0.667 n=2+2)
Enabled/csv/on            46.5B ±55%     21.0B ± 5%   ~     (p=0.667 n=2+2)
Enabled/csv/on-2          3.50B ±14%     2.00B ± 0%   ~     (p=0.333 n=2+2)
Enabled/csv/on-4         0.50B ±100%    0.50B ±100%   ~     (p=1.000 n=2+2)
Enabled/json2/50pct        104B ±32%      122B ±14%   ~     (p=0.667 n=2+2)
Enabled/json2/50pct-2     10.0B ± 0%      7.0B ±29%   ~     (p=0.333 n=2+2)
Enabled/json2/50pct-4     1.00B ± 0%     1.50B ±33%   ~     (p=1.000 n=2+2)
Enabled/json2/on          25.0B ± 4%     37.0B ± 3%   ~     (p=0.333 n=2+2)
Enabled/json2/on-2        1.50B ±33%     1.50B ±33%   ~     (p=1.000 n=2+2)
Enabled/json2/on-4        0.00B          0.00B        ~     (all equal)

name                   old allocs/op  new allocs/op  delta
Enabled/csv/50pct          0.00           0.00        ~     (all equal)
Enabled/csv/50pct-2        0.00           0.00        ~     (all equal)
Enabled/csv/50pct-4        0.00           0.00        ~     (all equal)
Enabled/csv/on             0.00           0.00        ~     (all equal)
Enabled/csv/on-2           0.00           0.00        ~     (all equal)
Enabled/csv/on-4           0.00           0.00        ~     (all equal)
Enabled/json2/50pct       0.50 ±100%      1.00 ± 0%   ~     (p=1.000 n=2+2)
Enabled/json2/50pct-2      0.00           0.00        ~     (all equal)
Enabled/json2/50pct-4      0.00           0.00        ~     (all equal)
Enabled/json2/on           0.00           0.00        ~     (all equal)
Enabled/json2/on-2         0.00           0.00        ~     (all equal)
Enabled/json2/on-4         0.00           0.00        ~     (all equal)
```
